### PR TITLE
Initial biscuit implementation

### PR DESCRIPTION
**Pull request by daeMOn63**
Brings bases for biscuit creation and validation

### CODE REVIEW

File: biscuit.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+	symbols.Extend(authority.symbols)
+
+	pbAuthority, err := proto.Marshal(tokenBlockToProtoBlock(authority))
+	if err != nil {
+		return nil, err
+	}
+
+	ts := &sig.TokenSignature{}
+	ts.Sign(rand.Reader, root, pbAuthority)
```
Comment (titanous) [inline biscuit.go:50]: This should use `rng`, which should default to `rand.Reader` if `nil`.

File: builder.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	b.caveats = append(b.caveats, &datalog.Caveat{Queries: []datalog.Rule{rule.convert(b.symbols)}})
+	return nil
+}
+
+func (b *builder) AddRight(resource, right string) error {
+	return b.AddAuthorityFact(&Fact{
+		Predicate: Predicate{
+			Name: "right",
+			IDs: []Atom{
+				Symbol("authority"),
```
Comment (titanous) [inline builder.go:83]: we should probably have a `const symAuthority` somewhere.

File: biscuit.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	blocks    []*Block
+	symbols   *datalog.SymbolTable
+	container *pb.Biscuit
+}
+
+var (
+	// ErrSymbolTableOverlap is returned when multiple blocks declare the same symbols
+	ErrSymbolTableOverlap = errors.New("symbol table overlap")
+	// ErrInvalidAuthorityIndex occurs when an authority block index is not 0
+	ErrInvalidAuthorityIndex = errors.New("invalid authority index")
```
Comment (titanous) [inline biscuit.go:30]: These error messages should have a `biscuit: ` prefix.

File: biscuit.go [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+func (b *Biscuit) Verify(root sig.PublicKey) (Verifier, error) {
+	if err := b.checkRootKey(root); err != nil {
+		return nil, err
+	}
+
+	return NewVerifier(b)
+}
+
+func (b *Biscuit) Caveats() [][]*datalog.Caveat {
+	result := append([][]*datalog.Caveat{}, b.authority.caveats)
```
Comment (titanous) [inline biscuit.go:152]: I think we should `make([][]*datalog.Caveat, 0, len(b.blocks)+1)` first.

File: biscuit.go [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+		result = append(result, block.caveats)
+	}
+	return result
+}
+
+func (b *Biscuit) Serialize() ([]byte, error) {
+	return proto.Marshal(b.container)
+}
+
+func (b *Biscuit) Print() string {
```
Comment (titanous) [inline biscuit.go:163]: Do you think this should be `String()`?

File: builder.go [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+
+		symbolsStart: symbols.Len(),
+		symbols:      symbols,
+		facts:        new(datalog.FactSet),
+	}
+}
+
+func (b *builder) AddAuthorityFact(fact *Fact) error {
+	if len(fact.Predicate.IDs) == 0 || fact.Predicate.IDs[0] != SymAuthority {
+		fact.Predicate.IDs[0] = SymAuthority
```
Comment (titanous) [inline builder.go:55]: This looks like it can panic if `IDs` hasn't been allocated?

File: builder.go [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+	if !b.facts.Insert(dlFact) {
+		return ErrDuplicateFact
+	}
+
+	return nil
+}
+
+func (b *builder) AddAuthorityRule(rule *Rule) error {
+	if len(rule.Head.IDs) == 0 || rule.Head.IDs[0] != SymAuthority {
+		rule.Head.IDs[0] = SymAuthority
```
Comment (titanous) [inline builder.go:68]: Ditto.

File: builder.go [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+		index:   0,
+		symbols: b.symbols.SplitOff(b.symbolsStart),
+		facts:   b.facts,
+		rules:   b.rules,
+		caveats: b.caveats,
+		context: b.context,
+	})
+}
+
+func From(serialized []byte) (*Biscuit, error) {
```
Comment (titanous) [inline builder.go:105]: I think I'd prefer calling this `Unmarshal` to match Go conventions.

File: builder.go [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+	return FromWithSymbols(serialized, DefaultSymbolTable)
+}
+
+func FromWithSymbols(serialized []byte, symbols *datalog.SymbolTable) (*Biscuit, error) {
```
Comment (titanous) [inline builder.go:109]: Do you think it makes more sense to have an `Unmarshaler` struct that has a `Symbols` field? That's a pretty standard Go convention that allows adding more options later.

File: datalog/datalog.go [hunk 10]
```diff
@@ -630,6 +639,50 @@ func (t *SymbolTable) Str(sym Symbol) string {
...
+// SplitOff returns a newly allocated slice containing the elements in the range
+// [at, len). After the call, the receiver will be left containing
+// the elements [0, at) with its previous capacity unchanged.
+func (t *SymbolTable) SplitOff(at int) *SymbolTable {
+	if at > len(*t) {
+		panic("split index out of bound")
+	}
+
+	new := (*t)[at:]
+	*t = (*t)[:at]
```
Comment (titanous) [inline datalog/datalog.go:656]: Is there any scenario where the returned `SymbolTable` and this "parent" will be both appended to later? If so, we need to copy as we could inadvertently modify another table. If not, add a comment noting that the returned value is intended to be immutable.

File: sig/crypto.go [hunk 11]
```diff
@@ -141,6 +141,36 @@ func (s *TokenSignature) Verify(pubkeys []PublicKey, msgs [][]byte) error {
 	return nil
 }
 
+func (s *TokenSignature) Serialize() ([][]byte, []byte) {
```
Comment (titanous) [inline sig/crypto.go:144]: I think this should be called `Encode`.

File: sig/crypto.go [hunk 12]
```diff
@@ -141,6 +141,36 @@ func (s *TokenSignature) Verify(pubkeys []PublicKey, msgs [][]byte) error {
...
+	params := make([][]byte, len(s.Params))
+	for i, p := range s.Params {
+		params[i] = p.Encode([]byte{})
+	}
+
+	return params, s.Z.Encode([]byte{})
+}
+
+func Deserialize(params [][]byte, z []byte) (*TokenSignature, error) {
```
Comment (titanous) [inline sig/crypto.go:153]: `Decode`?

File: sig/crypto.go [hunk 13]
```diff
@@ -141,6 +141,36 @@ func (s *TokenSignature) Verify(pubkeys []PublicKey, msgs [][]byte) error {
...
+	decodedParams := make([]*r255.Element, len(params))
+	for i, p := range params {
+		e := r255.NewElement()
```
Comment (titanous) [inline sig/crypto.go:156]: This should be `&r255.Element{}` to avoid unnecessarily zeroing first.

File: sig/crypto.go [hunk 14]
```diff
@@ -141,6 +141,36 @@ func (s *TokenSignature) Verify(pubkeys []PublicKey, msgs [][]byte) error {
...
+		if err := e.Decode(p); err != nil {
+			return nil, err
+		}
+		decodedParams[i] = e
+	}
+
+	decodedZ := r255.NewScalar()
```
Comment (titanous) [inline sig/crypto.go:163]: This should be `&r255.Scalar{}` to avoid unnecessarily zeroing first.

File: types.go [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+type Block struct {
+	index   uint32
+	symbols *datalog.SymbolTable
+	facts   *datalog.FactSet
+	rules   []*datalog.Rule
+	caveats []*datalog.Caveat
+	context string
+}
+
+func (b *Block) Print(symbols *datalog.SymbolTable) string {
```
Comment (titanous) [inline types.go:34]: `String()`?

File: types.go [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+const (
+	AtomTypeSymbol AtomType = iota
+	AtomTypeVariable
+	AtomTypeInteger
+	AtomTypeString
+	AtomTypeDate
+)
+
+const (
+	AuthoritySymbol = Symbol("authority")
```
Comment (titanous) [inline types.go:251]: This should be `SymbolAuthority`.

File: builder.go (hunk 9 above)
Comment (daeMOn63) [reply builder.go]: What do you think making it a `Marshaller`  and move the `Biscuit.Serialize`  on it too ?

https://github.com/flynn/biscuit-go/blob/1486c9c8266ff40dbb2a6f5cf48bedc569004af9/biscuit.go#L160-L162 

I like getting all the marshalling stuff on the same struct, but unsure with the `biscuit.container` private field in the way.

File: builder.go [hunk 17]
```diff
@@ -0,0 +1,247 @@
...
+		symbols:      symbols,
+		facts:        new(datalog.FactSet),
+	}
+}
+
+func (b *builder) AddAuthorityFact(fact *Fact) error {
+	if len(fact.Predicate.IDs) == 0 {
+		fact.Predicate.IDs = []Atom{SymbolAuthority}
+	} else if fact.Predicate.IDs[0] != SymbolAuthority {
+		fact.Predicate.IDs[0] = SymbolAuthority
```
Comment (titanous) [inline builder.go:57]: I think we need to make a copy of the fact in these cases. In cases where the exact same "template" fact is used across multiple builders, we can end up modifying data being used in multiple places, causing races.

File: builder.go [hunk 18]
```diff
@@ -0,0 +1,247 @@
...
+	}
+
+	return nil
+}
+
+func (b *builder) AddAuthorityRule(rule *Rule) error {
+	if len(rule.Head.IDs) == 0 {
+		rule.Head.IDs = []Atom{SymbolAuthority}
+	} else if rule.Head.IDs[0] != SymbolAuthority {
+		rule.Head.IDs[0] = SymbolAuthority
```
Comment (titanous) [inline builder.go:72]: Ditto, we should copy before modifying the rule.

File: builder.go [hunk 19]
```diff
@@ -0,0 +1,247 @@
...
+		caveats: b.caveats,
+		context: b.context,
+	})
+}
+
+type Unmarshaler struct {
+	Symbols *datalog.SymbolTable
+}
+
+var DefaultUnmarshaler = &Unmarshaler{Symbols: DefaultSymbolTable}
```
Comment (titanous) [inline builder.go:113]: I'd unexport this and make a top-level `Unmarshal` function, that way the symbol table doesn't inadvertently get modified globally.

File: builder.go [hunk 20]
```diff
@@ -0,0 +1,247 @@
...
+
+func (u *Unmarshaler) Unmarshal(serialized []byte) (*Biscuit, error) {
+	if u.Symbols == nil {
+		return nil, errors.New("biscuit: unmarshaler require a symbol table")
```
Comment (titanous) [inline builder.go:117]:
```suggestion
		return nil, errors.New("biscuit: unmarshaler requires a symbol table")
```

File: builder.go [hunk 21]
```diff
@@ -0,0 +1,247 @@
...
+	pubKeys := make([]sig.PublicKey, len(container.Keys))
+	for i, pk := range container.Keys {
+		pubKey, err := sig.NewPublicKey(pk)
+		if err != nil {
+			return nil, err
+		}
+		pubKeys[i] = pubKey
+	}
+
+	signedBlocks := [][]byte{container.Authority}
```
Comment (titanous) [inline builder.go:146]: You should allocate this as a slice with capacity `len(container.Blocks)+1` first.

**Pull request merged**